### PR TITLE
feat: add meta-learning example using protonets

### PR DESCRIPTION
**Pull request by liamcli**
## Description
Add meta-learning example that trains Protonets for few-shot learning on Omniglot.  See original paper for more details about the algorithm:
http://papers.nips.cc/paper/6996-prototypical-networks-for-few-shot-learning.pdf

This suggests that we should be able to implement MAML and Reptile with certain restrictions in our framework as well with a very nice way of parallelizing the inner loop by task.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: examples/experimental/omniglot_protonet_pytorch/20way1shot.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  embedding_dim: 64 # embedding number of channels
+
+resources:
+  slots_per_trial: 2
+
+searcher:
+  name: single
+  metric: loss
+  smaller_is_better: true
+
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/20way1shot.yaml:36]: remove space here

File: examples/experimental/omniglot_protonet_pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+## Prototypical Networks on Omniglot
+This example implements a meta-learning method called prototypical networks (ProtoNets) for few-shot learning on the Omniglot dataset.  ProtoNets often outperform MAML on few-shot learning problems and is one of the most popular methods for meta-learning.
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/README.md:2]: This example implements a meta-learning method... -> This example implements a popular meta-learning method. 

remove: "and is one of the most popular methods ...."

File: examples/experimental/omniglot_protonet_pytorch/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+### Folders and Files
+* **model_def.py**: Contains the core code for the model and protonet loss function following Determined's PyTorchTrial interface.
+* **data.py**: Contains the code used to generate few-shot learning tasks.
+* **const.yaml**: Contains the configuration for the experiment. 
+* **startup-hook.sh**: Script run before starting the trial that downloads the Omniglot data.
+* **fetch_data.sh**: Script to download the Omniglot data.
+
+### Data
+We use the Omniglot download script from the [meta-blocks](https://github.com/alshedivat/meta-blocks/blob/master/benchmarks/omniglot/fetch_data.sh) repo.  This script is called in `startup-hook.sh` so the data folder is available prior to starting the trial.
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/README.md:13]: how long does it take to download the data?

File: examples/experimental/omniglot_protonet_pytorch/data.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
+import os
+import numpy as np
+
+import torch
+from torch.utils.data import Dataset
+
+from PIL import Image
+
+# Each training example will be a classification task with n classes.
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/data.py:9]: nit: add this to a file comment at the top

File: examples/experimental/omniglot_protonet_pytorch/data.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    ):
+        """
+        Each subfolder of the parent data directory is a separate
+        alphabet with multiple characters.  Each class is a certain
+        character within an alphabet.  Tasks will correspond to 
+        n-way classification problems where the learner will 
+        have to predict which of n classes a test image belongs to.
+
+        Args:
+            # NOTE: You should partition all inds to a train set and a validation
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/data.py:32]: is inds indexes? maybe we should use idx. <- unless that's the OG example

File: examples/experimental/omniglot_protonet_pytorch/data.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                self.class_paths[class_idx] = [os.path.join(root, f) for f in files]
+                class_idx += 1
+
+        self.num_classes = num_classes
+        self.num_support = num_support
+        self.num_query = (
+            min_class_examples - num_support if (num_query is None) else num_query
+        )
+
+    def get_collate_fn(self):
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/data.py:59]: little confused here. Is this just using the default collate function? If so, we shouldn't need to define it.

File: examples/experimental/omniglot_protonet_pytorch/data.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        task_classes = np.random.choice(
+            self.class_inds, size=self.num_classes, replace=False
+        )
+        rotations = np.random.choice(
+            self.rotations, size=self.num_classes, replace=True
+        )
+
+        imgs = []
+        labels = []
+        for i, c in enumerate(task_classes):
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/data.py:78]: lets use better loops variables besides 'c'

File: examples/experimental/omniglot_protonet_pytorch/data.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            imgs_paths = np.random.choice(
+                self.class_paths[c],
+                size=(self.num_support + self.num_query),
+                replace=False,
+            )
+            for p in imgs_paths:
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/data.py:84]: same as above

File: examples/experimental/omniglot_protonet_pytorch/model_def.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+
+import numpy as np
+import determined as det
+from determined.pytorch import DataLoader, LRScheduler, PyTorchTrial, TorchData
+from data import OmniglotTasks
+
+
+# ProtoNet implementation based on https://github.com/jakesnell/prototypical-networks/blob/master/protonets/models/few_shot.py
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/model_def.py:13]: Same as above. Move this into a file comment at the top

File: examples/experimental/omniglot_protonet_pytorch/model_def.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+
+    def optimizer(self, model: nn.Module):
+        optimizer = torch.optim.Adam(
+            model.parameters(),
+            lr=self.context.get_hparam("learning_rate"),
+            weight_decay=self.context.get_hparam("weight_decay"),
+        )
+        return optimizer
+
+    def create_lr_scheduler(self, optimizer: torch.optim.Optimizer):
```
Comment (katport) [inline examples/experimental/omniglot_protonet_pytorch/model_def.py:185]: move optimizer and create_lr_scheduler above loss(). hard to look back and forth for referencing.

File: examples/experimental/omniglot_protonet_pytorch/README.md (hunk 3 above)
Comment (liamcli) [reply examples/experimental/omniglot_protonet_pytorch/README.md:13]: data download is really fast, less than 15 seconds.

File: examples/experimental/omniglot_protonet_pytorch/data.py (hunk 6 above)
Comment (liamcli) [reply examples/experimental/omniglot_protonet_pytorch/data.py:59]: this is different from the default collate_fn.  default zips dictionary field values and returns a list for a given field whereas I am creating a list of dictionaries.

Comment (katport) [reply examples/experimental/omniglot_protonet_pytorch/data.py:59]: Let's add some comments then describing this.

Comment (liamcli) [reply examples/experimental/omniglot_protonet_pytorch/data.py:59]: added comment
